### PR TITLE
High level Swarm network simulation

### PR DESCRIPTION
**Pull request by janos**
This PR includes TestSwarmNetwork test and changes to the swarm that are required for that test to pass. There are minor lint warning changes to make travis happy.

TestSwarmNetwork contains tests for static network simulation with 10 and 100 nodes, and also tests where nodes are added or removed while the availability of files is checked between changes. On every test step, a number of nodes is created and the same number of files are uploaded on each of nodes (one file per node). The check consists of trying to retrieve all files from all nodes.

To run the standard set of tests:

```
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork
```

To run long running tests, too:

```
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork -longrunning
```

To run specific tests from the table:

```
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/10_nodes
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/100_nodes -longrunning
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/100_nodes -longrunning
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/inc_node_count -longrunning
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/dec_node_count -longrunning
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/dec_inc_node_count
go test -v -count 1 github.com/ethereum/go-ethereum/swarm -run TestSwarmNetwork/inc_dec_node_count -longrunning
```

Every test execution can include `-waitkademlia` cli flag to wait that Kademlia table on every node is healthy before files retrieval is checked.

### CODE REVIEW

File: swarm/network/stream/messages.go [hunk 1]
```diff
@@ -238,15 +240,18 @@ func (p *Peer) handleOfferedHashesMsg(req *OfferedHashesMsg) error {
 	}
 	go func() {
 		select {
-		case <-time.After(600 * time.Second):
-			p.Drop(errors.New("handle offered hashes timeout"))
+		case <-time.After(120 * time.Second):
+			log.Warn("ERROR in handleOfferedHashesMsg, DROPPING peer!", "err", "TIMEOUT")
+			p.Drop(err)
```
Comment (zelig) [inline swarm/network/stream/messages.go]: err is what? why change timeout msg?

File: swarm/network/stream/stream.go [hunk 2]
```diff
@@ -600,7 +601,11 @@ func (c *client) batchDone(p *Peer, req *OfferedHashesMsg, hashes []byte) error
 }
 
 func (c *client) close() {
-	close(c.next)
+	select {
```
Comment (zelig) [inline swarm/network/stream/stream.go:604]: not safe against concurrent calls

File: swarm/network_test.go [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+	var errCount int
+	for err := range errc {
+		if err != nil {
+			errCount++
+		}
+		log.Error(err.Error())
+	}
+
+	log.Info("check stats", "total check count", totalCheckCount, "total files found", atomic.LoadUint64(&totalFoundCount), "total errors", errCount)
```
Comment (zelig) [inline swarm/network_test.go]: why need atomic at this point?

File: swarm/storage/memstore.go [hunk 4]
```diff
@@ -186,11 +186,9 @@ func (s *MemStore) Put(entry *Chunk) {
 			if entry.ReqC == nil {
 				entry.ReqC = node.entry.ReqC
 			}
-			// if node.entry.dbStoredC != nil {
-			// 	entry.dbStored = node.entry.dbStored
-			// 	entry.dbStoredC = node.entry.dbStoredC
-			// 	entry.dbStoredMu = node.entry.dbStoredMu
-			// }
+			entry.dbStored = node.entry.dbStored
```
Comment (zelig) [inline swarm/storage/memstore.go]: lets please get rid of this lock and bool. it was a measure to handle a bug. 
we should not write in a batch twice in dbstore thats all

File: swarm/network/stream/stream.go (hunk 2 above)
Comment (janos) [reply swarm/network/stream/stream.go:604]: yes, but client.close() is called in peer.removeClient() which is under a lock, would protect this select? https://github.com/ethersphere/go-ethereum/blob/5347db84e2ca5d53b3ab9cd6077e02f67decaee8/swarm/network/stream/peer.go#L279

File: swarm/network/stream/delivery.go [hunk 5]
```diff
@@ -136,19 +137,23 @@ func (d *Delivery) handleRetrieveRequestMsg(sp *Peer, req *RetrieveRequestMsg) e
 	chunk, created := d.db.GetOrCreateRequest(req.Key)
 	if chunk.ReqC != nil {
 		if created {
-			if err := d.RequestFromPeers(chunk.Key[:], false, sp.ID()); err != nil {
+			if err := d.RequestFromPeers(chunk.Key[:], true, sp.ID()); err != nil {
```
Comment (gbalint) [inline swarm/network/stream/delivery.go:140]: Can we really always SkipCheck when we forward a request?

**Review by zelig — APPROVED** [review]

File: swarm/network/kademlia.go [hunk 6]
```diff
@@ -206,6 +206,14 @@ func (k *Kademlia) Register(peers []OverlayAddr) error {
 		k.addrCountC <- k.addrs.Size()
 	}
 	// log.Trace(fmt.Sprintf("%x registered %v peers, %v known, total: %v", k.BaseAddr()[:4], size, known, k.addrs.Size()))
+
+	if k.nDepthC != nil {
```
Comment (holisticode) [inline swarm/network/kademlia.go]: Can we please have more comments? What is this check doing? Just helps reading the code faster once one has to look back later (or for new developers)

File: swarm/network/kademlia.go [hunk 7]
```diff
@@ -369,6 +377,13 @@ func (k *Kademlia) Off(p OverlayConn) {
 		if k.addrCountC != nil {
 			k.addrCountC <- k.addrs.Size()
 		}
+		if k.nDepthC != nil {
```
Comment (holisticode) [inline swarm/network/kademlia.go]: Just so I understand, why do we not need any `NotifyDepth` after this check?

File: swarm/network/stream/syncer.go [hunk 8]
```diff
@@ -188,16 +201,16 @@ func NewSwarmSyncerClient(_ *Peer, db *storage.DBAPI, ignoreExistingRequest bool
...
 	})
 }
 
 // NeedData
 func (s *SwarmSyncerClient) NeedData(key []byte) (wait func()) {
-	chunk, created := s.db.GetOrCreateRequest(key)
+	chunk, _ := s.db.GetOrCreateRequest(key)
 	// TODO: we may want to request from this peer anyway even if the request exists
-	if chunk.ReqC == nil || (s.ignoreExistingRequest && !created) {
+	if chunk.ReqC == nil { //|| (s.ignoreExistingRequest && !created) {
```
Comment (holisticode) [inline swarm/network/stream/syncer.go]: can we already say if the commented part can definitively be removed now?

File: swarm/network_test.go [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+
+	swarms := make(map[discover.NodeID]*Swarm)
+	files := make([]file, 0)
+
+	services := map[string]adapters.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext) (node.Service, error) {
+			config := api.NewConfig()
+			config.PssEnabled = false
+
+			dir, err := ioutil.TempDir(dir, "node")
```
Comment (holisticode) [inline swarm/network_test.go:216]: shouldn't there be a `defer os.RemoveAll(dir)`?

File: swarm/network/stream/stream.go (hunk 2 above)
Comment (janos) [reply swarm/network/stream/stream.go:604]: I will trace why removeClient would be called multiple times, it should not happen for sure.

File: swarm/network/kademlia.go (hunk 6 above)
Comment (janos) [reply swarm/network/kademlia.go]: :+1: Will add more comments. There are some on nDepthC and NeighbourhoodDepthC, but they may be far away from this part of the code.

File: swarm/network/kademlia.go (hunk 7 above)
Comment (janos) [reply swarm/network/kademlia.go]: I think that they are serving a different purpose. NotifyDepth is notifying peers, but this channel is sending depth to some other part of the code for the same node. And we use it in the stream package for updating sync subscriptions. This is one of the purposes that TestSwarmNetwork is testing when adding and removing nodes.

File: swarm/network/stream/syncer.go (hunk 8 above)
Comment (janos) [reply swarm/network/stream/syncer.go]: We need recheck ignore existing request functionality, as it creates problems with the test. For now, we decided to remove the functionality. Will add some comments around it.

File: swarm/network_test.go (hunk 9 above)
Comment (janos) [reply swarm/network_test.go:216]: Actually, this is by design not needed. In order to avoid tracking and removing all nodes directories, they are all under one temp dir which is created here https://github.com/ethersphere/go-ethereum/pull/352/files/159007b743b062e317806d757e220285920e2303#diff-8135ad1202bf3e545e65dafb7002da9cR194, and removed at the end of the test.

File: swarm/storage/ldbstore.go [hunk 10]
```diff
@@ -512,11 +516,13 @@ func (s *LDBStore) Put(chunk *Chunk) {
 		go func() {
 			<-batchC
 			chunk.markAsStored()
+			log.Info("ldbstore.put success (go)", "key", chunk.Key)
```
[the remark this answers is not in the document]
Comment (janos) [reply swarm/storage/ldbstore.go]: 👍

File: cmd/swarm/db.go [hunk 11]
```diff
@@ -115,5 +115,6 @@ func openLDBStore(path string, basekey []byte) (*storage.LDBStore, error) {
 
 	storeparams := storage.NewDefaultStoreParams()
 	ldbparams := storage.NewLDBStoreParams(storeparams, path)
+	ldbparams.BaseKey = basekey
```
[the remark this answers is not in the document]
Comment (gbalint) [reply cmd/swarm/db.go:118]: This looks like patching to me, it should be nicer to not set the basekey after constructing, but add it as a constructor parameter

File: swarm/api/config.go [hunk 12]
```diff
@@ -113,13 +115,15 @@ func (self *Config) Init(prvKey *ecdsa.PrivateKey) {
...
 	self.BzzKey = keyhex
+	self.NodeID = discover.PubkeyID(&prvKey.PublicKey).String()
 
 	if self.SwapEnabled {
 		self.Swap.Init(self.Contract, prvKey)
 	}
 
 	self.privateKey = prvKey
 	self.LocalStoreParams.Init(self.Path)
+	self.LocalStoreParams.BaseKey = common.FromHex(keyhex)
```
Comment (gbalint) [inline swarm/api/config.go:126]: Same here, we should not patch the params retroactively imho.

File: swarm/network/stream/delivery.go [hunk 13]
```diff
@@ -208,14 +213,13 @@ R:
...
 		d.db.Put(chunk)
-		chunk.WaitToStore()
-		err = chunk.GetErrored()
-		if err != nil {
+
+		go func(req *ChunkDeliveryMsg) {
+			err := chunk.WaitToStore()
 			if err == storage.ErrChunkInvalid {
 				req.peer.Drop(err)
 			}
```
Comment (gbalint) [inline swarm/network/stream/delivery.go:221]: Maybe I miss something, but we don't we have to close chun.ReqC here?

File: swarm/network_test.go [hunk 14]
```diff
@@ -0,0 +1,577 @@
...
+	rand.Seed(time.Now().UnixNano())
+
+	flag.Parse()
+
+	log.Root().SetHandler(log.LvlFilterHandler(log.Lvl(*loglevel), log.StreamHandler(colorable.NewColorableStderr(), log.TerminalFormat(true))))
+}
+
+// TestSwarmNetwork runs a series of test simulations with
+// static and dynamic Swarm nodes in network simulation, by
+// uploading files to every node and retreiving them.
```
Comment (nonsense) [inline swarm/network_test.go]: typo - retrieving.

File: swarm/network_test.go (hunk 3 above)
Comment (janos) [reply swarm/network_test.go]: I had some bad value repostings without it. Maybe atomic can be replaced with a locked int as a counter. That would be more idiomatic and maybe even correct?

File: cmd/swarm/db.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (janos) [reply cmd/swarm/db.go:118]: I agree, this was the fix with the least invasive approach. We should revaluate storage types parametrization.

File: swarm/storage/ldbstore.go [hunk 15]
```diff
@@ -227,7 +227,11 @@ func encodeIndex(index *dpaDBIndex) []byte {
 }
 
 func encodeData(chunk *Chunk) []byte {
-	return append(chunk.Key[:], chunk.SData...)
+	// Always create a new underlying array for the returned byte slice.
+	// The chunk.Key array may be used in the returned slice which
+	// may be changed later in the code or by the LevelDB, resulting
+	// that the Key is changed as well.
+	return append(append([]byte{}, chunk.Key[:]...), chunk.SData...)
```
Comment (nonsense) [inline swarm/storage/ldbstore.go:234]: @janos I trust you now what you are doing, but why is this necessary? Aren't we copying everything returned by LevelDB into its own memory, before assigning it to `chunk.Key` and `chunk.SData`? How can LevelDB be changing our chunks after they've been retrieved?

File: swarm/network/stream/delivery.go (hunk 13 above)
Comment (janos) [reply swarm/network/stream/delivery.go:221]: ReqC closing is actually moved to LocalStore.Put.

File: swarm/storage/localstore.go [hunk 16]
```diff
@@ -97,26 +97,52 @@ func (self *LocalStore) Put(chunk *Chunk) {
 	}
```
Comment (nonsense) [inline swarm/storage/localstore.go:109]: A bit of godoc for `Put` and its semantics would be appreciated :)

File: swarm/swarm.go [hunk 17]
```diff
@@ -129,10 +128,18 @@ func NewSwarm(ctx *node.ServiceContext, backend chequebook.Backend, config *api.
...
-	nodeid := discover.PubkeyID(crypto.ToECDSAPub(common.FromHex(config.PublicKey)))
-	addr := network.NewAddrFromNodeID(nodeid)
+
+	nodeID, err := discover.HexID(config.NodeID)
+	if err != nil {
+		return nil, err
+	}
+	addr := &network.BzzAddr{
+		OAddr: common.FromHex(config.BzzKey),
+		UAddr: []byte(discover.NewNode(nodeID, net.IP{127, 0, 0, 1}, 30303, 30303).String()),
```
Comment (nonsense) [inline swarm/swarm.go:138]: We should make this configurable at some point.

**Review by nonsense — APPROVED** [review]
lgtm, great work on the tests!

I have added a few minor comment. Along them:

1. `stream/stream.go` could use some godoc
2. `stream/delivery.go` could use some godoc

File: cmd/swarm/db.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (nonsense) [reply cmd/swarm/db.go:118]: I agree we should re-evaluate the storage types parameters - I came across some weird instances where we were passing way too many parameters to constructors that didn't care about them, which is not great.

File: swarm/storage/ldbstore.go (hunk 15 above)
Comment (janos) [reply swarm/storage/ldbstore.go:234]: That may look like, for sure. But, the slice created from the key array may contain the same array, and when it is used in append, the returned slice may reference the same array, or the append can create a new array and copy all bytes into it. If the new array is created, then there in no problem, but if the the same array is used, it reference will be passed to the leveldb, allowing it to modify the slice, actually the shared underlaying array.

File: swarm/swarm.go (hunk 17 above)
Comment (janos) [reply swarm/swarm.go:138]: 👍

**Review by gbalint — APPROVED** [review]

File: swarm/storage/ldbstore.go (hunk 15 above)
Comment (nonsense) [reply swarm/storage/ldbstore.go:234]: @janos yes, this is clear. I understand why you are doing this. I guess I am missing where we are using the `encodeFunc`, anyway, it looks fine.

File: swarm/storage/localstore.go (hunk 16 above)
Comment (janos) [reply swarm/storage/localstore.go:109]: Yes, it contains some important logic. Updated.

File: swarm/storage/ldbstore.go (hunk 15 above)
Comment (janos) [reply swarm/storage/ldbstore.go:234]: Yes, `encodeFunc` is used as `LDBStore.doPut` function as the `LDBStore.encodeDataFunc` to generate the storage data saved under `dkey`. It was needed to be parametrized in LDBStore, for replacing it with the mock store encode data function.
